app/assets/javascripts/components/AccountMenu.tsx

```diff
@@ -0,0 +1,984 @@
+          month: 'short',
+          hour: 'numeric',
+          minute: 'numeric'
+        });
+      }
+
+      return f.format(protectionExpiry);
+    }
+    return null;
+  };
+
+  const passcodeInputRef = useRef<HTMLInputElement>();
+  const emailInputRef = useRef<HTMLInputElement>();
+  const passwordInputRef = useRef<HTMLInputElement>();
+
+  const [showLogin, setShowLogin] = useState(false);
+  const [showRegister, setShowRegister] = useState(false);
+  const [showAdvanced, setShowAdvanced] = useState(false);
+  const [isAuthenticating, setIsAuthenticating] = useState(false);
+  const [email, setEmail] = useState('');
+  const [password, setPassword] = useState('');
+  const [passwordConfirmation, setPasswordConfirmation] = useState<string | undefined>(undefined);
+  const [status, setStatus] = useState<string | undefined>(undefined);
+  const [syncError, setSyncError] = useState<string | undefined>(undefined);
+
+  const [isEphemeral, setIsEphemeral] = useState(false);
+  const [isStrictSignIn, setIsStrictSignIn] = useState(false);
+
+  const [passcode, setPasscode] = useState<string | undefined>(undefined);
+  const [passcodeConfirmation, setPasscodeConfirmation] = useState<string | undefined>(undefined);
+
+  const [encryptionStatusString, setEncryptionStatusString] = useState<string | undefined>(undefined);
+  const [isEncryptionEnabled, setIsEncryptionEnabled] = useState(false);
+  const [shouldMergeLocal, setShouldMergeLocal] = useState(true);
+
+  const [server, setServer] = useState<string | undefined>(undefined);
+  const [url, setUrl] = useState<string | undefined>(undefined);
+  const [showPasscodeForm, setShowPasscodeForm] = useState(false);
+  const [selectedAutoLockInterval, setSelectedAutoLockInterval] = useState<unknown>(null);
+  const [isImportDataLoading, setIsImportDataLoading] = useState(false);
+  const [isErrorReportingEnabled, setIsErrorReportingEnabled] = useState(false);
+  const [appVersion, setAppVersion] = useState('');
+  const [hasPasscode, setHasPasscode] = useState(application.hasPasscode());
+  const [isBackupEncrypted, setIsBackupEncrypted] = useState(isEncryptionEnabled);
+  const [isSyncInProgress, setIsSyncInProgress] = useState(false);
+  const [protectionsDisabledUntil, setProtectionsDisabledUntil] = useState(getProtectionsDisabledUntil());
+
+  const reloadAutoLockInterval = useCallback(async () => {
+    const interval = await application.getAutolockService().getAutoLockInterval();
+    setSelectedAutoLockInterval(interval);
+  }, [application]);
+
+  const user = application.getUser();
+  const errorReportingIdValue = errorReportingId();
+  const canAddPasscode = !application.isEphemeralSession();
+  const keyStorageInfo = StringUtils.keyStorageInfo(application);
+  const passcodeAutoLockOptions = application.getAutolockService().getAutoLockIntervalOptions();
+  const showBetaWarning = appState.showBetaWarning;
+
+  const focusWithTimeout = (inputElementRef: RefObject<HTMLInputElement>) => {
```

Comment:
I'm thinking it may be a little bit cleaner to have an emailInputFocused state, and check it on a useEffect hook. We'd do setEmailInputFocused(true) when we want the input to be focused on the next render, and then on the useEffect hook, something like this:

```
useEffect(() => {
  if (emailInputFocused) {
    setEmailInputFocused(false);
    inputElementRef.current?.focus();
  }
}
```

Reply:
It's a very good idea, done :)